hax/hax/motr/__init__.py

```diff
@@ -148,52 +162,93 @@ def _entrypoint_request_cb(self, reply_context: Any, req_id: Any,
             LOG.exception('Failed to notify failure for %s', process_fid)
 
         LOG.debug('enqueue entrypoint request for %s', remote_rpc_endpoint)
-        self.queue.put(
-            EntrypointRequest(reply_context=reply_context,
-                              req_id=req_id,
-                              remote_rpc_endpoint=remote_rpc_endpoint,
-                              process_fid=process_fid,
-                              git_rev=git_rev,
-                              pid=pid,
-                              is_first_request=is_first_request))
+        entrypoint_req: EntrypointRequest = EntrypointRequest(
+            reply_context=reply_context,
+            req_id=req_id,
+            remote_rpc_endpoint=remote_rpc_endpoint,
+            process_fid=process_fid,
+            git_rev=git_rev,
+            pid=pid,
+            is_first_request=is_first_request)
+
+        # If rconfc from motr land sends an entrypoint request when
+        # the hax consumer thread is already stopping, there's no
+        # point in en-queueing the request as there's no one to process
+        # the same. Thus, invoke send_entrypoint_reply directly.
+        self.send_entrypoint_request_reply(entrypoint_req)
 
     def send_entrypoint_request_reply(self, message: EntrypointRequest):
         reply_context = message.reply_context
         req_id = message.req_id
         remote_rpc_endpoint = message.remote_rpc_endpoint
         process_fid = message.process_fid
+        e_rc = EAGAIN
 
         LOG.debug('Processing entrypoint request from remote endpoint'
                   " '{}', process fid {}".format(remote_rpc_endpoint,
                                                  str(process_fid)))
         sess = principal_rm = confds = None
         try:
             util = self.consul_util
-            sess = util.get_leader_session_no_wait()
-            principal_rm = util.get_session_node(sess)
-            confds = util.get_confd_list()
-            rm_fid = util.get_rm_fid()
+            # When stopping, there's a possibility that hax may receive
+            # an entrypoint request from motr land. In order to unblock
+            # motr land, reply with entrypoint request with no confds
+            # and RM endpoints as the processes might have already
+            # stopped.
+            rc_quorum = 0
+            if self.is_stopping:
+                confds = []
+                rm_fid = Fid(0, 0)
```

Comment:
Why isn't it enough to do this only? Why do we need some additional `self.consul.failed_svcs` list?

Reply:
So, even before shutdown starts (i.e. self.is_stopping = True), as hax notifies about confd failure (STOPPED), rconfc from hax motr land starts re-election and RM starts re-connecting, new entrypoint request is received before shutdown starts and hax then replies with a non-empty confd list and rm endpoint as `self.is_stopping` is still false. On confd failure, rconfc stops and starts again and this goes on in a loop,
```
static void rconfc_start(struct m0_rconfc *rconfc)
{
        struct m0_ha                   *ha  = m0_get()->i_ha;
        struct m0_ha_entrypoint_client *ecl = &ha->h_entrypoint_client;
        int                             rc  = 0;

        M0_PRE(rconfc->rc_local_conf == NULL);

        M0_ENTRY();

        while (rconfc_state(rconfc) != M0_RCS_FAILURE) {
                rconfc_state_set(rconfc, M0_RCS_ENTRYPOINT_WAIT);

                if (rconfc->rc_ha_entrypoint_rep.hae_control == M0_HA_ENTRYPOINT_QUERY) {
                        M0_LOG(M0_DEBUG, "Querying ENTRYPOINT...");
                        m0_ha_entrypoint_client_request(ecl);
                        break;
                }

                M0_LOG(M0_DEBUG, "ENTRYPOINT ready...");
                rc = rconfc_start_internal(rconfc);
                rconfc->rc_ha_entrypoint_rep.hae_control = M0_HA_ENTRYPOINT_QUERY;
                if (rc == 0)
                        break;
                m0_ha_entrypoint_rep_free(&rconfc->rc_ha_entrypoint_rep);
                M0_LOG(M0_DEBUG, "Try reconnecting after rc=%d", rc);
                M0_CNT_INC(rconfc->rc_ha_entrypoint_retries);
        }
```
This blocks rconfc shutdown and hax shutdown. In order to break this, we need to reply with only alive confds else empty.

Reply:
This does not sound like a robust solution for breaking this. Because during the cluster shutdown confd may be already down or shutting down while hax still will think it's alive...

Reply:
Sorry didn't get you, we also invoke m0_rconfc_stop() and m0_rconfc_fini() through m0_spiel_rconfc_stop().

Reply:
Imagine confd is stopping. hax' rconfc starts re-election in Motr land. But `self.consul.failed_svcs` is not updated yet - because the notification did not come yet. hax will reply with non-empty confd in entry-point reply, but there is no confd any more. Is such a scenario possible?

Reply:
Makes sense now?

Reply:
Not maintaining failed_svc anymore, but fetching statuses from Consul. So now if the confd status in Consul is not STARTED then we do not include that confd in the entrypoint reply.

Reply:
Don't you think the same situation may happen also in this case? Do we have a STOPPING state for Motr processes in Consul KV?

Reply:
Yes, Every state reported by a motr process is updated in consul kv.

Reply:
Then it should be fine, I suppose...